File: ios/src/main/java/y2k/joyreactor/platform/StoryboardNavigation.java
package y2k.joyreactor.platform;

import org.robovm.apple.foundation.NSArray;
import org.robovm.apple.foundation.NSMutableArray;
import org.robovm.apple.foundation.NSURL;
import org.robovm.apple.uikit.UIApplication;
import org.robovm.apple.uikit.UINavigationController;
import org.robovm.apple.uikit.UIViewController;
import y2k.joyreactor.ImageViewController;
import y2k.joyreactor.Post;

/**
 * Created by y2k on 02/10/15.
 */
public class StoryboardNavigation implements Navigation {

    private static Post sPostArgument; // TODO:
    private static String sPostIdArgument; // TODO:

    @Override
    public void switchProfileToLogin() {
        switchTo("Login");
    }

    @Override
    public void switchLoginToProfile() {
        switchTo("Profile");
    }

    @Override
    public void closeCreateComment() {
        getNavigationController().popViewController(true);
    }

    @Override
    public void closeAddTag() {
        getNavigationController().popViewController(true);
    }

    @Override
    public void openPost(String postId) {
        sPostIdArgument = postId;
        getNavigationController().pushViewController(instantiateViewController("Post"), true);
    }

    @Override
    public String getArgumentPostId() {
        return sPostIdArgument;
    }

    private void switchTo(String storyboardId) {
        NSArray<UIViewController> stack = new NSMutableArray<>(getNavigationController().getViewControllers());
        stack.remove(stack.size() - 1);
        stack.add(instantiateViewController(storyboardId));
        getNavigationController().setViewControllers(stack, true);
    }

    private UIViewController instantiateViewController(String storyboardId) {
        return getNavigationController()
                .getStoryboard()
                .instantiateViewController(storyboardId);
    }

    public static UINavigationController getNavigationController() {
        return (UINavigationController) UIApplication
                .getSharedApplication()
                .getWindows().get(0)
                .getRootViewController();
    }

    @Override
    public void openBrowser(String url) {
        UIApplication.getSharedApplication().openURL(new NSURL(url));
    }

    @Override
    public void openVideo(String postId) {
        sPostIdArgument = postId;
        getNavigationController().pushViewController(instantiateViewController("Video"), true);
    }

    @Override
    public void openImageView(Post post) {
        sPostArgument = post;
        getNavigationController().pushViewController(new ImageViewController(), true);
    }

    @Override
    public void openCreateComment() {
        // TODO:
    }

    @Override
    public void openPostGallery() {
        // TODO:
    }
}

File: ios/src/main/kotlin/y2k/joyreactor/platform/StoryboardNavigation.kt
package y2k.joyreactor.platform

import org.robovm.apple.foundation.NSMutableArray
import org.robovm.apple.foundation.NSURL
import org.robovm.apple.uikit.UIApplication
import org.robovm.apple.uikit.UINavigationController
import org.robovm.apple.uikit.UIViewController
import kotlin.reflect.KClass

/**
 * Created by y2k on 02/10/15.
 */
class StoryboardNavigation : NavigationService {

    override fun close() {
        throw UnsupportedOperationException()
    }

    override fun <T : Any> open(vmType: KClass<T>, argument: String) {
        throw UnsupportedOperationException()
    }

    //    override fun openMessages(name: String) {
//        sArgument = name
//        navigationController.pushViewController(instantiateViewController("Messages"), true)
//    }
//
//    override fun switchProfileToLogin() {
//        switchTo("Login")
//    }
//
//    override fun switchLoginToProfile() {
//        switchTo("Profile")
//    }
//
//    override fun closeCreateComment() {
//        navigationController.popViewController(true)
//    }
//
//    override fun closeAddTag() {
//        navigationController.popViewController(true)
//    }
//
//    override fun openPost(postId: String) {
//        sArgument = postId
//        navigationController.pushViewController(instantiateViewController("Post"), true)
//    }

    override val argument: String
        get() = sArgument!!

    private fun switchTo(storyboardId: String) {
        val stack = NSMutableArray(navigationController.viewControllers.toList())
        stack.removeAt(stack.size - 1)
        stack.add(instantiateViewController(storyboardId))
        navigationController.setViewControllers(stack, true)
    }

    private fun instantiateViewController(storyboardId: String): UIViewController {
        return navigationController.storyboard.instantiateViewController(storyboardId)
    }

    override fun openBrowser(url: String) {
        UIApplication.getSharedApplication().openURL(NSURL(url))
    }

//    override fun openVideo(postId: String) {
//        sArgument = postId
//        navigationController.pushViewController(instantiateViewController("Video"), true)
//    }
//
//    override fun openImageView(postId: String) {
//        sArgument = postId
//        navigationController.pushViewController(ImageViewController(), true)
//    }
//
//    override fun openCreateComment() {
//        // TODO:
//    }
//
//    override fun openPostGallery() {
//        // TODO:
//    }

    companion object {

        private var sArgument: String? = null // TODO:

        val navigationController: UINavigationController
            get() = UIApplication.getSharedApplication().windows[0].rootViewController as UINavigationController
    }
}